### PR TITLE
Introduce a common SpecMapper

### PR DESCRIPTION
**Pull request by kanitw**
~~Merge #4528 first.~~

Part of #4376

### CODE REVIEW

File: src/normalize/base.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+  S extends GenericSpec<any, any>, // Input type
+  O extends NormalizedSpec, // Output Type
+  SN extends GenericSpec<any, any> = S // input to additional normalization
+> {
+  name: string;
+  hasMatchingType: (spec: GenericSpec<any, any>, config: Config) => spec is S;
+
+  run(spec: S, params: NormalizerParams, normalize: Normalize<SN, O>): O;
+
+  // TODO: add config cleaner
```
Comment (domoritz) [inline src/normalize/base.ts]: Is there an issue for this? What does this even mean?

File: src/normalize/core.ts [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+    // Special handling for a faceted unit spec as it can return a facet spec, not just a layer or unit spec like a normal unit spec.
+    if (isUnitSpec(spec)) {
+      const hasRow = channelHasField(spec.encoding, ROW);
+      const hasColumn = channelHasField(spec.encoding, COLUMN);
+
+      if (hasRow || hasColumn) {
+        return this.mapFacetedUnit(spec, params);
+      }
+    }
+    // TODO: add extra type here
```
Comment (domoritz) [inline src/normalize/core.ts]: Issue?

File: src/normalize/base.ts (hunk 1 above)
Comment (kanitw) [reply src/normalize/base.ts]: I'll remove this comment

**Comment by kanitw** [comment]
Removed comments, added comment to https://github.com/vega/vega-lite/issues/4376 instead.

**Pull request merged**